File: bms/go/geo/geo_r3_PreciseVector_Dot/geo_r3_PreciseVector_Dot.go
package test

import "math/big"

var ()

const (
	geo_r3_prec = big.MaxPrec
)

func (v geo_r3_PreciseVector) Dot(ov geo_r3_PreciseVector) *big.Float {
	return geo_r3_precAdd(geo_r3_precMul(v.X, ov.X), geo_r3_precAdd(geo_r3_precMul(v.Y, ov.Y), geo_r3_precMul(v.Z, ov.Z)))
}

func geo_r3_precMul(a, b *big.Float) *big.Float {
	return new(big.Float).SetPrec(geo_r3_prec).Mul(a, b)
}

func geo_r3_precAdd(a, b *big.Float) *big.Float {
	return new(big.Float).SetPrec(geo_r3_prec).Add(a, b)
}

type geo_r3_PreciseVector struct {
	X, Y, Z *big.Float
}

File: translations_only/fail/go/geo/geo_r3_PreciseVector_Dot/gemini/BaseRepair/geo_r3_PreciseVector_Dot.rs

use std::ops::{Add, Mul};

const GEO_R3_PREC: usize = 1024;

fn dot(v: &geo_r3_PreciseVector, ov: &geo_r3_PreciseVector) -> f64 {
    geo_r3_prec_add(
        &geo_r3_prec_mul(&v.x, &ov.x),
        &geo_r3_prec_add(
            &geo_r3_prec_mul(&v.y, &ov.y),
            &geo_r3_prec_mul(&v.z, &ov.z),
        ),
    )
}

fn geo_r3_prec_mul(a: &f64, b: &f64) -> f64 {
    a * b
}

fn geo_r3_prec_add(a: &f64, b: &f64) -> f64 {
    a + b
}

struct geo_r3_PreciseVector {
    x: f64,
    y: f64,
    z: f64,
}
